pageserver/src/tenant/timeline.rs

```diff
@@ -4176,6 +4177,398 @@ impl Timeline {
+            let layers = tokio::select! {
+                guard = ancestor.layers.read() => guard,
+                _ = self.cancel.cancelled() => {
+                    return Err(ShuttingDown);
+                }
+                _ = ancestor.cancel.cancelled() => {
+                    return Err(ShuttingDown);
+                }
+            };
+
+            let (later_by_lsn, straddling, rest) =
+                detach_ancestor::partition_work(ancestor_lsn, &layers);
+
+            // between retries, these can change if compaction or gc ran in between. this will mean
+            // we have to redo work.
+            (later_by_lsn, straddling, rest)
+        };
+
+        // TODO: layers are already sorted by something: use that to determine how much of remote
+        // copies are already done.
+        tracing::info!(filtered=%filtered_layers, to_rewrite = straddling_branchpoint.len(), historic=%rest_of_historic.len(), "collected layers");
+
+        // TODO: copying and lsn prefix copying could be done at the same time with a single fsync after
+        let mut new_layers: Vec<Layer> =
+            Vec::with_capacity(straddling_branchpoint.len() + rest_of_historic.len());
+
+        {
+            tracing::debug!(to_rewrite = %straddling_branchpoint.len(), "copying prefix of delta layers");
+
+            let mut tasks = tokio::task::JoinSet::new();
+
+            let mut wrote_any = false;
+
+            let limiter = Arc::new(tokio::sync::Semaphore::new(
+                options.rewrite_concurrency.get(),
+            ));
+
+            for layer in straddling_branchpoint {
+                let limiter = limiter.clone();
+                let timeline = self.clone();
+                let ctx = ctx.detached_child(TaskKind::DetachAncestor, DownloadBehavior::Download);
+
+                tasks.spawn(async move {
+                    let _permit = limiter.acquire().await;
+                    let copied = detach_ancestor::upload_rewritten_layer(
+                        end_lsn,
+                        &layer,
+                        &timeline,
+                        &timeline.cancel,
+                        &ctx,
+                    )
+                    .await?;
+                    Ok(copied)
+                });
+            }
+
+            while let Some(res) = tasks.join_next().await {
+                match res {
+                    Ok(Ok(Some(copied))) => {
+                        wrote_any = true;
+                        tracing::info!(layer=%copied, "rewrote and uploaded");
+                        new_layers.push(copied);
+                    }
+                    Ok(Ok(None)) => {}
+                    Ok(Err(e)) => return Err(e),
+                    Err(je) => return Err(Unexpected(je.into())),
+                }
+            }
+
+            // FIXME: the fsync should be mandatory, after both rewrites and copies in the case of
+            // failpoint usage
+            if wrote_any {
```

Comment:
`wrote_any` is equivalent to `!new_layers.is_empty()` , let's get rid of it.

Reply:
It is currently when there's a:
1. copy_lsn_prefix
2. copy historic (which does not hardlink at the moment)

Not doing the change.